### PR TITLE
WIP: Add logr.WithCallDepth()

### LINKED ISSUE #30: call stack skipping
When wrapping a logr instance with some additional helper functions, it may be necessary to inform the wrapped instance that it has to skip additional levels in the call stack to find the actual source code location for the log message.

### LINKED ISSUE #32: Generic way to add options without breaking changes
This was originally a comment on PR #31 that implements the idea proposed in #30.

There will always be a demand for more options to the Logger interface, but we have to balance the added value of these new options with the impact on the ecosystem of breaking changes. Constant breaking changes to the interface make this project less attractive. Ideally we will be able to define a v1.0 that will remain stable for years to come, without too many breaking changes getting there.

I would like to find a generic way to add arbitrary options and then just break compatibility once to add this new interface.

For example, we could add this method to the interface:

```go
WithOption(option Option) Logger
```

with `Option` being some interface. This could be nothing more than an interface with some marker method that never actually gets called:

```go
type Option interface {
    LogrOption()
}
```

Then we could define a new `CallDepth` option, which you would use as follows:

```go
log = log.WithOption(logr.CallDepth(2))
```

Implementations would be required to ignore unknown options. Applications would be able to define their own custom options that get passed around.

We would always be able to add new 'official' options to the logr package without breaking the interface. 

If we also add something similar to convert the logger into something else, we may never have to break the Logger interface again:

```go
Convert(Conversion) interface{}
```

Usage example that would solve #11:

```go
underlying := log.Convert(logr.Underlying).(*logrus.Logger)
```

### PR DESCRIPTION
**Pull request by thockin**
Fixes #30
Fixes #32

### CODE REVIEW

File: logr.go [hunk 1]
```diff
@@ -188,6 +188,14 @@ type Logger interface {
...
 	// (see the package documentation for more information).
 	WithName(name string) Logger
+
+	// WithCallDepth returns a Logger that will offset the call-stack by the
+	// specified number of frames.  This is useful for users who have helper
+	// functions between the "useful" call site and the actual calls to Logger
+	// methods.  Implementations that log information about the call site (such
+	// as file, function, or line) would otherwise log information about the
+	// intermediate helper functions.  Successive calls to this are additive.
+	WithCallDepth(depth int) Logger
```
Comment (pohly) [inline logr.go:198]: We need to document how depth is counted.

"With 0 depth, the default, the source code location of the direct caller of the logging functions is recorded. A non-zero depth skips that number of additional call stacks."

Or something like that...

**Comment by wojas** [comment]
If we add a new method to the interface, it will break existing code if even one of its dependencies upgrades to the newer version. I understand that this is still not 1.0, but the standard for breaking everything should be a bit higher for an interface that an entire ecosystem depends on.

I suggest adding this (and any future additions) as a new interface that any consumer can specifically test for, e.g.:

``go
if wcd, ok := log.(logr.WithCallDepther); ok {
    log = wcd.WithCallDepth(3)
} 
```

We can provide convenience functions to make this easier:

```go
log = logr.WithCallDepth(log, 2)
```

This would call the new method if the logger supports it, or just return the exiting logger if not.

The implementation of this function could be something like:

```go
func WithCallDepth(log Logger, depth int) Logger {
    if wcd, ok := log.(logr.WithCallDepther); ok {
        log = wcd.WithCallDepth(depth)
    } 
    return log
}
```


The Go filesystem proposal does something similar for the same reason:
https://go.googlesource.com/proposal/+/master/design/draft-iofs.md#extensions-to-fs

**Comment by pohly** [comment]
> If we add a new method to the interface, it will break existing code if even one of its dependencies upgrades to the newer version. I understand that this is still not 1.0, but the standard for breaking everything should be a bit higher for an interface that an entire ecosystem depends on.

How many implementations of the logr interface exist at the moment? I agree that it breaks users of those unless those implementations get updated first, but perhaps that is a manageable size?

Another option to ease the pain for users of an unmodified implementation might be to have a helper in logr:
```
// Deprecated, will be removed.
type LoggerV03 interface {
	Enabled() bool
	Info(msg string, keysAndValues ...interface{})
	Error(err error, msg string, keysAndValues ...interface{})
	V(level int) Logger
	WithValues(keysAndValues ...interface{}) Logger
	WithName(name string) Logger
}

// This is the interface that all implementations should implement.
type Logger interface {
        LoggerV03
	WithCallDepth(depth int) Logger
}

// UpdateLoggerV03 turns an implementation of the v0.3.0 Logger interface into
// an implementation of the current interface.
func UpdateLoggerV03(l LoggerV03) Logger {
    return updateLogger{LoggerV03: l}
}

type updateLogger struct {
    LoggerV03
}

func (l updateLogger) WithCallDepth(depth int) Logger {
    return l
}
````

**Comment by wojas** [comment]
I don't know how big the impact is, but do consider that projects that are using it are likely to have their own logr implementations used for testing to to capture some logs for an internal subsystem in a special way.

Constant breaking changes to the interface make this project less attractive.

I suppose there will always be a demand for more adding more options to the interface, but let's find a generic way to add arbitrary options and then just break compatibility once to add this interface.

For example, we could add this to the interface:

```go
WithOption(option Option) Logger
```

with `Option` being some interface.

Then we could define a new  `CallDepth` option, which you would use as follows:

```go
log = log.WithOption(logr.CallDepth(2))
```

Implementations would be required to ignore unknown options. Applications would be able to define their own custom options that get passed around.

If we also add something similar to convert the logger into something else, we may never have to break the Logger interface again:

```go
Convert(Conversion) interface{}
```

Usage example:

```go
underlying := log.Convert(logr.Underlying).(*logrus.Logger)
```

**Comment by wojas** [comment]
I will create a new issue with the Option proposal and can create a PR for if we can agree on it.

**Comment by thockin** [comment]
@wojas 

> the standard for breaking everything should be a bit higher for an interface that an entire ecosystem depends on.

I'm not sure that an entire ecosystem depends on this, and if that were true, perhaps we really should call it 1.0.  The point of being pre 1.0 is to get feedback like this and not feel TOO bad about making API changes.

> `log = logr.WithCallDepth(log, 2)`
>
> This would call the new method if the logger supports it, or just return the exiting logger if not.

So the question I have is whether callers who SPECIFICALLY invoke this logic really want the implementation to ignore their request or whether they would rather compile fail and then think "oh right, I need to update logrus"?

I understand the ripple effect of that.  I updated my kubernetes client, that uses logr.Logger.WithCallDepth.  So now I need to update my logrus dep, how annoying.  Oh, well that needs some other lib update.  Forget it. I just won't update.

The flip side is "all my kubernetes client logs are showing the wrong caller info, and I only found that out because I am debugging a critical outage and they lead me down a dead end, FML".

I know I can waive this away now as a not-yet-1.0 problem, but it will certainly arise later.

**Comment by miquella** [comment]
Although thought is being put into the level of effort to migrate the surrounding ecosystem to an updated interface, I believe a more important thought mentioned earlier is being eclipsed.

@wojas mentioned the design around extensions to the upcoming [FS package](https://tip.golang.org/pkg/io/fs/):

> The Go filesystem proposal does something similar for the same reason:
> https://go.googlesource.com/proposal/+/master/design/draft-iofs.md#extensions-to-fs

Features that were not requisite for implementations (read as: optional features) were specifically excluded from the core interface. Instead, supplementary interfaces are used to provide these features, often with a paired helper method or two for ergonomics.

Similarly, stack traces are specifically called out in this package as an optional and implementation-dependent feature. So the suggestion to move this method out of the core interface and into an optional interface, possibly with a helper method or two, seems very reasonable.

Multiple helper functions could be provided with different semantics. For example: one that silently ignores the call if the underlying logger doesn't implement the optional interface and one that errors. These helpers could even be moved into a sub-package to keep the core package clean and simple.

My two cents. I'm an outsider to the project, so please take it with a grain of salt. :slightly_smiling_face:

**Comment by thockin** [comment]
I appreciate all reasonable input, so thanks.  This got longer than I expected.
Summary at bottom.

Indeed this is not a problem unique to logr and there are varying degrees of impact in the case of not-supported features.  You called out stack trace - fair example.

Would we add `.WithStackTrace()` to the Logger interface?  It seems pretty similar in form to `.WithCallDepth()`, so I suppose the same reasoning would apply.

>  Multiple helper functions could be provided with different semantics

Indeed.  My main concern is not what we could offer, but how much these things will ACTUALLY be used.  For such a low-level library, every new func/method/type we expose carries a cost.  If the preponderance of opinion is that something won't actually be used, then we shouldn't add it until that shifts.

I also ponder how much we should encapsulate and how much we should just expose and let users deal with.  E.g. the FS proposal seems to assume that users can deal with testing for a `ReadFileFS` interface.

E.g.

```
func LogHelper(logger logr.Logger, msg string) {
    // Sets call depth if supported, pass-thru otherwise
    logr.WithCallDepth(logger, 1).Info(msg)
}
```

vs.

```
func LogHelper(logger logr.Logger, msg string) {
    // Sets call depth if supported, nil otherwise
    // Presumably we'd ALSO offer the above pass-thru option
    ld := logr.WithCallDepthOrNil(logger, 1)
    if ld == nil {
        ld = logger.WithValues("caller", "unknown")
    }
    ld.Info(msg)
}
```

vs.

```
func LogHelper(logger logr.Logger, msg string) {
    // Set call depth if supported, 
    if ld, ok := logger.(logr.CallDepthLogger); ok
        logger = ld.WithCallDepth(1)
    } else {
        logger = logger.WithValues("caller", "unknown")
    }
    logger.Info(msg)
}
```

It's sligthly more LOC at the call-site, but it leaves the user to decide what
to do.  To publish the set of optional interfaces that implementations can
support, we'd need to expose those interface types anyway.  This presumes that
the number of such call-sites is relatively low, and for log helpers that would
be true.  For call-stacks, maybe not. That would probably encourage users to
write wrappers.

E.g.

```
type myLogger struct {
    logr.Logger
}

// NewLogger returns a logger that will include a stack-trace on *all* Error() calls.
func NewLogger() logr.Logger {
    logger := newRealLogger() // glogr, zapr, whatever
    if stackLogger, ok := logger.(logr.CallStackLogger); ok {
        logger = stackLogger.WithCallStack(logr.CallStackError)
    } else {
        // Would be nice to get some impl-name to log here.  Maybe repeat the
        // above pattern for logr.Metadata.
        logger.Info("call-stack logging is not supported")
    }
    return &myLogger{logger}
}

func DoRealWork(logger logr.Logger) {
    // ...
    // This log-line will include a stack trace
    logger.Error(err, "something bad")
    // ...
}
```

vs.

```
// MyLogger is a utility interface that makes it easier for my own callers to
// log a stack trace, but can still be used in places that expect log.Logger.
type MyLogger interface {
    logr.Logger
    // Let the caller decide to log stack-trace or not.
    WithStackTrace() logr.Logger
}

type myLogger struct {
    logr.Logger
}

func NewLogger() MyLogger {
    logger := newRealLogger() // glogr, zapr, whatever
    return &myLogger{logger}
}

func (ml *myLogger) WithStackTrace() logr.Logger {
    if stackLogger, ok := ml.(logr.CallStackLogger); ok {
        // Only log stack on Error() calls
        logger := stackLogger.WithCallStack(logr.CallStackAlways)
        return &myLogger{logger}
    }
    return ml
}

func DoRealWork(logger MyLogger) {
    // ...
    // This log-line will include a stack trace if supported
    logger.WithStackTrace().Error(err, "something bad")
    // ...
}
```

vs.

```
// Assume go-logr/logr exposes helper methods and just the Logger type.

func DoRealWork(logger MyLogger) {
    // ...
    // This log-line will include a stack trace if supported
    logr.WithStackTrace(logger).Error(err, "something bad")
    // ...
}
```

So I guess the question is how much to encapsulate into a (small) set of helper
functions per concept vs just exposing alternate Interfaces and letting callers
do it.  If we think almost all projects will end up writing the same wrappers,
we should probably own that.

Secondary to that, how parameterizable those helper functions would need be.
E.g.

```
// Always include stack trace if supported
func WithStackTrace() Logger {}
```

vs

```
// Include stack trace if supported and policy enables it.
func WithStackTrace(when StackTracePolicy) Logger {}

type StackTracePolicy int

const (
    StackTraceAlways StackTracePolicy = iota
    StackTraceError  StackTracePolicy
)
```

vs

```
// Always include stack trace
func WithStackTrace() Logger {}

// Include stack trace on errors
func WithStackTraceOnError() Logger {}
```

If you join that with the `...OrNil` pattern you end up with 2x as many helper
functions.  That said, if we have to expose the interfaces, maybe the implicit
"or pass-through" helper functions are sufficient.  Anyone who wants
`...OrNil()` can use the interface assertions.

> These helpers could even be moved into a sub-package to keep the core package clean and simple.

That's an interesting idea. What naming convention would make this palatable?


# Summary

If we want to start using a future-safe pattern for optional extensions, I
think I have talked myself into:

```
// in logr

type CallDepthLogger interface {
    Logger
    WithCallDepth(depth int)
}

func WithCallDepth(logger Logger, depth int) Logger {
    if decorator, ok := logger.(CallDepthLogger); ok {
        return decorator.WithCallDepth(depth)
    }
    return logger
}

type CallStackLogger interface {
    Logger
    WithCallStack()
    WithCallStackOnError()
}

func WithCallStack(logger Logger) Logger {
    if decorator, ok := logger.(CallStackLogger); ok {
        return decorator.WithCallStack()
    }
    return logger
}

func WithCallStackOnError(logger Logger) Logger {
    if decorator, ok := logger.(CallStackLogger); ok {
        return decorator.WithCallStackOnError()
    }
    return logger
}
```

**Comment by wojas** [comment]
@thockin I like the approach proposed under Summary. In the end the filesystem pattern is less clunky than #32 and avoids breaking compatibility now and in the future. Just having this pattern available as a way to add new features will probably also allow releasing a v1.0 sooner rather than later, without fear that we may have forgotten something important.

If a consumer for some reason really cares if the logger actually implements a certain extension, it can always directly test against the extension interface.

>> These helpers could even be moved into a sub-package to keep the core package clean and simple.
> That's an interesting idea. What naming convention would make this palatable?

After reconsidering it, I would prefer to just keep these in the main package. This makes the extensions more easily discoverable, easier to use and I do not expect too many of them anyway.

**Comment by miquella** [comment]
> > These helpers could even be moved into a sub-package to keep the core package clean and simple.
> 
> That's an interesting idea. What naming convention would make this palatable?

My thoughts around this idea aren't very well developed, but the initial idea was in one of two forms:

1. Host the primary interfaces and functions in the root package and all extension interfaces in a sub-package: `logrextensions` or similar.
2. Host all interfaces and functions related to a particular extension (e.g. stack traces) together in their own package: `stacktraces` or similar.

Personally, unless the extensions are plentiful or complex, I find option (1) far more palatable. Having the extensions separate from the main package keeps both packages focused and minimal. However, as @wojas pointed out, any option where extensions are separated from the main package does decrease their discoverability.

----

I suppose it's a balance between discoverability of extensions and the extra noise in the main package.

If the extensions aren't readily discoverable, they won't be used, but if there is too much noise in the main package, it becomes more difficult for new users to understand how to adopt (e.g. "which pieces do I actually need to make this thing work?).

With a very small number of easy to understand extensions, it's not difficult. The challenge often comes when the number of extensions or their complexity grows.

**Comment by wojas** [comment]
> 1. Host the primary interfaces and functions in the root package and all extension interfaces in a sub-package: logrextensions or similar.

Unfortunately this is not possible, because it would create an import loop (helper func imports extension interface, which imports logr.Logger).

> With a very small number of easy to understand extensions, it's not difficult. The challenge often comes when the number of extensions or their complexity grows.

Note that if someone comes up with complex extensions that kind of stretch what the scope of the logr interface is, it is always possible to maintain such extensions outside of the logr repository.

**Comment by miquella** [comment]
> > 1. Host the primary interfaces and functions in the root package and all extension interfaces in a sub-package: logrextensions or similar.
> 
> Unfortunately this is not possible, because it would create an import loop (helper func imports extension interface, which imports logr.Logger).

That would be true if the extension helper functions were hosted in the main package instead of the sub-package. However, if the helper functions are hosted along side the extension interface in the sub-package, the main package wouldn't have a reason to import the sub-package, so there wouldn't be any import cycles.

**Comment by miquella** [comment]
> > With a very small number of easy to understand extensions, it's not difficult. The challenge often comes when the number of extensions or their complexity grows.
> 
> Note that if someone comes up with complex extensions that kind of stretch what the scope of the logr interface is, it is always possible to maintain such extensions outside of the logr repository.

My thinking was along these lines when I originally suggested splitting the extension pieces into separate package(s).

**Comment by thockin** [comment]
I'm not sure yet that moving to a sub-package is necessary.  I'll update this PR and solicit feedback on comments and docs.

**Comment by miquella** [comment]
FWIW, I agree. Splitting out a new package for a single interface and helper method seems premature at this point.

**Review by pohly — COMMENTED** [review]
Do we need to document the meaning of the `depth` parameter or is is clear enough already that `depth = 0` means "skip no additional stack frames"?

Other than that this looks okay to me.

**Comment by thockin** [comment]
I clarified comments a bit.

**Review by pohly — APPROVED** [review]

**Review by DirectXMan12 — APPROVED** [review]
Seems pretty reasonable to me!  I like this approach. :+1:

**Pull request merged**